clients/src/main/resources/common/message/README.md

```diff
@@ -79,9 +87,14 @@ There are several primitive field types available.
 
 * "int64": a 64-bit integer.  This takes up 8 bytes on the wire.
 
-* "string": a string.  This must be less than 64kb in size when serialized as UTF-8.  This takes up 2 bytes on the wire, plus the length of the string when serialized to UTF-8.
+* "string": a string.  Strings are serialized as a length followed by the
+  contents as UTF-8.  The contents must be less than 64kb in size.  In
+  non-flexible versions, the string length will always be 2 bytes.  In flexible
+  versions, the length is a variable-length integer. 
 
-* "bytes": binary data.  This takes up 4 bytes on the wire, plus the length of the bytes.
+* "bytes": binary data.  Bytes fields are serialized as a length followed by
+  the contents.  In non-flexible versions, the length will always be 4 bytes.
+  In flexible versions, the length is a variable-length integer. 
 
 In addition to these primitive field types, there is also an array type.  Array
 types start with a "[]" and end with the name of the element type.  For
```

Comment:
This file doesn't yet mention it, but I think it should include a section on arrays, especially so with this PR. Prior, all arrays are either non-nullable or nullable with the same length encoding as bytes. There is one special array with varint encoding for number of headers in a record (but that has a separate definition, I guess). With flexible versions, arrays are varint encoded.

Reply:
Also maybe a note in the nullable section that nullable strings, arrays, bytes are encoded with a length of -1.

Reply:
When I originally wrote this, I didn't intend this section as a complete description of how the encoding worked, but just a general hint to how much space various datatypes took, so that people could choose the right ones.  Maybe we should just include a link to the serialization documentation here, though, to avoid repeating that same information in two places and/or getting out of date.

Reply:
I was referring to https://kafka.apache.org/protocol.html .  There will be additional material there due to COMPACT_STRING, COMPACT_ARRAY, etc.  There should probably be something there about tagged fields as well.  Let's do that in a follow-on PR.